gapis/gfxapi/sync/sync.go

```diff
@@ -96,3 +99,31 @@ func MutationAtomsFor(ctx context.Context, c *path.Capture, atoms *atom.List, id
 	transforms.Transform(ctx, *atoms, w)
 	return &a, nil
 }
+
+// MutateWithSubcommands returns a list of atoms that represent the correct
+// mutations to have the state for all atoms before and including the given index.
+func MutateWithSubcommands(ctx context.Context, c *path.Capture, atoms atom.List, callback func(state *gfxapi.State, subcommandIndex []uint64, a atom.Atom)) error {
+	// This is where we want to handle sub-states
+	// This involves transforming the tree for the given Indices, and
+	//   then mutating that.
+	rc, err := capture.ResolveFromPath(ctx, c)
+	if err != nil {
+		return err
+	}
+	s := rc.NewState()
+
+	for id, a := range atoms.Atoms {
+		if sync, ok := a.API().(SynchronizedAPI); ok {
+			if err := sync.MutateSubcommands(ctx, a, atom.ID(id), s, callback); err != nil && err == context.Canceled {
+				return err
+			}
+		} else {
+			if err := a.Mutate(ctx, s, nil); err != nil && err == context.Canceled {
+				return err
+			}
+		}
+		callback(s, []uint64{uint64(id)}, a)
```

Comment:
Although it may not hurt for calling `callback` twice when the atom's API is indeed a `SynchronizedAPI`, still prefer to move this second call to `callback` to the `else` branch. Or maybe adding some comments?